dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewScope.kt

```diff
@@ -0,0 +1,249 @@
+    }
+
+    // region RumScope
+
+    override fun handleEvent(
+        event: RumRawEvent,
+        writer: Writer<RumEvent>
+    ): RumScope? {
+
+        when (event) {
+            is RumRawEvent.SentError -> {
+                errorCount++
+                sendViewUpdate(writer)
+            }
+            is RumRawEvent.SentResource -> {
+                resourceCount++
+                sendViewUpdate(writer)
+            }
+            is RumRawEvent.SentAction -> {
+                actionCount++
+                sendViewUpdate(writer)
+            }
+            is RumRawEvent.StartView -> onStartView(event, writer)
+            is RumRawEvent.StopView -> onStopView(event, writer)
+            is RumRawEvent.StartAction -> onStartAction(event, writer)
+            is RumRawEvent.StartResource -> onStartResource(event, writer)
+            is RumRawEvent.AddError -> onAddError(event, writer)
+            else -> delegateEventToChildren(event, writer)
+        }
+
+        return if (stopped && activeResourceScopes.isEmpty()) {
+            null
+        } else {
+            this
+        }
+    }
+
+    override fun getRumContext(): RumContext {
+        return parentScope.getRumContext().copy(viewId = viewId)
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun onStartView(
+        event: RumRawEvent.StartView,
+        writer: Writer<RumEvent>
+    ) {
+        if (!stopped) {
+            sendViewUpdate(writer)
+            delegateEventToChildren(event, writer)
+            stopped = true
+        }
+    }
+
+    private fun onStopView(
+        event: RumRawEvent.StopView,
+        writer: Writer<RumEvent>
+    ) {
+        delegateEventToChildren(event, writer)
+        val startedKey = keyRef.get()
+        val shouldStop = (event.key == startedKey) || (startedKey == null)
+        if (shouldStop && !stopped) {
+            attributes.putAll(event.attributes)
+            sendViewUpdate(writer)
+            stopped = true
+        }
+    }
+
+    private fun onStartAction(
+        event: RumRawEvent.StartAction,
+        writer: Writer<RumEvent>
+    ) {
+        delegateEventToChildren(event, writer)
+        if (stopped) return
+
+        if (activeActionScope == null) {
+            activeActionScope = RumActionScope.fromEvent(this, event)
+        }
+    }
+
+    private fun onStartResource(
+        event: RumRawEvent.StartResource,
+        writer: Writer<RumEvent>
+    ) {
+        delegateEventToChildren(event, writer)
+        if (stopped) return
+
+        val updatedEvent = event.copy(attributes = updateAttributesWithActionId(event.attributes))
+        activeResourceScopes.put(
+            WeakReference(event.key),
+            RumResourceScope.fromEvent(this, updatedEvent)
+        )
+    }
+
+    private fun onAddError(
+        event: RumRawEvent.AddError,
+        writer: Writer<RumEvent>
+    ) {
+        delegateEventToChildren(event, writer)
```

Comment:
this code is repeated in each function...maybe you can extract it in a function that takes a lambda ...something like `delegateAndExecute(()->Unit)`